### PR TITLE
Support Photon OS autoinstall

### PR DESCRIPTION
**Pull request by keirazhang**
Signed-off-by: Qi Zhang <qiz@vmware.com>

### CODE REVIEW

**Comment by keirazhang** [comment]
Pipeline ID: 634
```
+----------------------------------------------------------------+
| Guest OS Type             | VMware Photon OS 3.0 x86_64        |
+----------------------------------------------------------------+
| VMTools Version           | 11.3.5.31214 (build-18557794)      |
+----------------------------------------------------------------+
| CloudInit Version         | 19.4                               |
+----------------------------------------------------------------+
| Config Guest Id           | vmwarePhoton64Guest                |
+----------------------------------------------------------------+
| Hardware Version          | vmx-20                             |
+----------------------------------------------------------------+
| GuestInfo Guest Id        | vmwarePhoton64Guest                |
+----------------------------------------------------------------+
| GuestInfo Guest Full Name | VMware Photon OS (64-bit)          |
+----------------------------------------------------------------+
| GuestInfo Guest Family    | linuxGuest                         |
+----------------------------------------------------------------+
| GuestInfo Detailed Data   | architecture='X86'                 |
|                           | bitness='64'                       |
|                           | distroName='VMware Photon OS'      |
|                           | distroVersion='3.0'                |
|                           | familyName='Linux'                 |
|                           | kernelVersion='4.19.132-5.ph3-esx' |
|                           | prettyName='VMware Photon OS 3.0'  |
+----------------------------------------------------------------+


Test Results (Total: 28, Failed: 0, No Run: 0, Elapsed Time: 01:55:23):
+-----------------------------------------------------------+
| Name                                 | Status | Exec Time |
+-----------------------------------------------------------+
| deploy_vm_efi_nvme_vmxnet3           | Passed | 00:09:59  |
| ovt_verify_install                   | Passed | 00:10:44  |
| ovt_verify_status                    | Passed | 00:00:40  |
| vgauth_check_service                 | Passed | 00:00:35  |
| check_ip_address                     | Passed | 00:00:41  |
| stat_balloon                         | Passed | 00:00:34  |
| stat_hosttime                        | Passed | 00:00:32  |
| check_inbox_driver                   | Passed | 00:00:47  |
| check_os_fullname                    | Passed | 00:00:31  |
| check_efi_firmware                   | Passed | 00:00:31  |
| cpu_hot_add_basic                    | Passed | 00:04:18  |
| check_quiesce_snapshot_custom_script | Passed | 00:01:19  |
| memory_hot_add_basic                 | Passed | 00:05:01  |
| device_list                          | Passed | 00:02:22  |
| secureboot_enable_disable            | Passed | 00:03:52  |
| e1000e_network_device_ops            | Passed | 00:01:50  |
| vmxnet3_network_device_ops           | Passed | 00:01:51  |
| cpu_multicores_per_socket            | Passed | 00:08:30  |
| gosc_perl_dhcp                       | Passed | 00:04:51  |
| gosc_perl_staticip                   | Passed | 00:04:37  |
| gosc_cloudinit_dhcp                  | Passed | 00:05:11  |
| gosc_cloudinit_staticip              | Passed | 00:04:52  |
| paravirtual_vhba_device_ops          | Passed | 00:01:55  |
| lsilogic_vhba_device_ops             | Passed | 00:02:14  |
| lsilogicsas_vhba_device_ops          | Passed | 00:01:55  |
| sata_vhba_device_ops                 | Passed | 00:02:14  |
| nvme_vhba_device_ops                 | Passed | 00:01:56  |
| ovt_verify_uninstall                 | Passed | 00:02:50  |
+-----------------------------------------------------------+
```

```
+---------------------------------------------------------------+
| Guest OS Type             | VMware Photon OS 4.0 x86_64       |
+---------------------------------------------------------------+
| VMTools Version           | 11.3.5.31214 (build-18557794)     |
+---------------------------------------------------------------+
| CloudInit Version         | 21.3                              |
+---------------------------------------------------------------+
| Config Guest Id           | vmwarePhoton64Guest               |
+---------------------------------------------------------------+
| Hardware Version          | vmx-20                            |
+---------------------------------------------------------------+
| GuestInfo Guest Id        | vmwarePhoton64Guest               |
+---------------------------------------------------------------+
| GuestInfo Guest Full Name | VMware Photon OS (64-bit)         |
+---------------------------------------------------------------+
| GuestInfo Guest Family    | linuxGuest                        |
+---------------------------------------------------------------+
| GuestInfo Detailed Data   | architecture='X86'                |
|                           | bitness='64'                      |
|                           | distroName='VMware Photon OS'     |
|                           | distroVersion='4.0'               |
|                           | familyName='Linux'                |
|                           | kernelVersion='5.10.61-1.ph4-esx' |
|                           | prettyName='VMware Photon OS 4.0' |
+---------------------------------------------------------------+


Test Results (Total: 28, Failed: 0, No Run: 0, Elapsed Time: 01:44:37):
+-----------------------------------------------------------+
| Name                                 | Status | Exec Time |
+-----------------------------------------------------------+
| deploy_vm_efi_paravirtual_vmxnet3    | Passed | 00:05:02  |
| ovt_verify_install                   | Passed | 00:06:20  |
| ovt_verify_status                    | Passed | 00:01:06  |
| vgauth_check_service                 | Passed | 00:00:54  |
| check_ip_address                     | Passed | 00:00:57  |
| stat_balloon                         | Passed | 00:00:50  |
| stat_hosttime                        | Passed | 00:00:51  |
| check_inbox_driver                   | Passed | 00:01:07  |
| check_os_fullname                    | Passed | 00:00:51  |
| check_efi_firmware                   | Passed | 00:00:53  |
| cpu_hot_add_basic                    | Passed | 00:04:29  |
| check_quiesce_snapshot_custom_script | Passed | 00:01:46  |
| memory_hot_add_basic                 | Passed | 00:05:20  |
| device_list                          | Passed | 00:02:38  |
| secureboot_enable_disable            | Passed | 00:04:10  |
| e1000e_network_device_ops            | Passed | 00:01:18  |
| vmxnet3_network_device_ops           | Passed | 00:01:21  |
| cpu_multicores_per_socket            | Passed | 00:08:21  |
| gosc_perl_dhcp                       | Passed | 00:05:12  |
| gosc_perl_staticip                   | Passed | 00:05:01  |
| gosc_cloudinit_dhcp                  | Passed | 00:05:53  |
| gosc_cloudinit_staticip              | Passed | 00:05:33  |
| paravirtual_vhba_device_ops          | Passed | 00:02:13  |
| lsilogic_vhba_device_ops             | Passed | 00:02:45  |
| lsilogicsas_vhba_device_ops          | Passed | 00:02:15  |
| sata_vhba_device_ops                 | Passed | 00:02:43  |
| nvme_vhba_device_ops                 | Passed | 00:02:10  |
| ovt_verify_uninstall                 | Passed | 00:02:38  |
+-----------------------------------------------------------+
```
```
+-------------------------------------------------------------------------------+
| Guest OS Type             | RedHat 8.5 x86_64                                 |
+-------------------------------------------------------------------------------+
| VMTools Version           | 11.2.5.26209 (build-17337674)                     |
+-------------------------------------------------------------------------------+
| CloudInit Version         | 21.1                                              |
+-------------------------------------------------------------------------------+
| Config Guest Id           | rhel8_64Guest                                     |
+-------------------------------------------------------------------------------+
| Hardware Version          | vmx-20                                            |
+-------------------------------------------------------------------------------+
| GuestInfo Guest Id        | rhel8_64Guest                                     |
+-------------------------------------------------------------------------------+
| GuestInfo Guest Full Name | Red Hat Enterprise Linux 8 (64-bit)               |
+-------------------------------------------------------------------------------+
| GuestInfo Guest Family    | linuxGuest                                        |
+-------------------------------------------------------------------------------+
| GuestInfo Detailed Data   | bitness='64'                                      |
|                           | distroName='Red Hat Enterprise Linux'             |
|                           | distroVersion='8.5'                               |
|                           | familyName='Linux'                                |
|                           | kernelVersion='4.18.0-348.el8.x86_64'             |
|                           | prettyName='Red Hat Enterprise Linux 8.5 (Ootpa)' |
+-------------------------------------------------------------------------------+


Test Results (Total: 28, Failed: 0, No Run: 2, Elapsed Time: 02:01:50):
+-------------------------------------------------------------+
| Name                                 |   Status | Exec Time |
+-------------------------------------------------------------+
| deploy_vm_efi_nvme_vmxnet3           |   Passed | 00:22:31  |
| ovt_verify_install                   |   Passed | 00:06:57  |
| ovt_verify_status                    |   Passed | 00:02:50  |
| vgauth_check_service                 |   Passed | 00:00:57  |
| check_ip_address                     |   Passed | 00:00:46  |
| stat_balloon                         |   Passed | 00:00:45  |
| stat_hosttime                        |   Passed | 00:00:43  |
| check_inbox_driver                   |   Passed | 00:01:02  |
| check_os_fullname                    |   Passed | 00:00:48  |
| check_efi_firmware                   |   Passed | 00:00:47  |
| cpu_hot_add_basic                    |   Passed | 00:04:56  |
| check_quiesce_snapshot_custom_script |   Passed | 00:01:30  |
| memory_hot_add_basic                 |   Passed | 00:05:17  |
| device_list                          |   Passed | 00:02:33  |
| secureboot_enable_disable            |   Passed | 00:04:09  |
| e1000e_network_device_ops            |   Passed | 00:01:22  |
| vmxnet3_network_device_ops           |   Passed | 00:01:29  |
| cpu_multicores_per_socket            |   Passed | 00:09:49  |
| gosc_perl_dhcp                       |   Passed | 00:04:38  |
| gosc_perl_staticip                   |   Passed | 00:04:22  |
| gosc_cloudinit_dhcp                  |   Passed | 00:04:45  |
| gosc_cloudinit_staticip              |   Passed | 00:04:35  |
| paravirtual_vhba_device_ops          |   Passed | 00:03:02  |
| lsilogic_vhba_device_ops             | * No Run | 00:00:45  |
| lsilogicsas_vhba_device_ops          | * No Run | 00:00:39  |
| sata_vhba_device_ops                 |   Passed | 00:03:05  |
| nvme_vhba_device_ops                 |   Passed | 00:02:48  |
| ovt_verify_uninstall                 |   Passed | 00:02:40  |
+-------------------------------------------------------------+
```
```
+-------------------------------------------------------------------------------+
| Guest OS Type             | SLED 15.3 x86_64                                  |
+-------------------------------------------------------------------------------+
| VMTools Version           | 11.2.5.26209 (build-17337674)                     |
+-------------------------------------------------------------------------------+
| CloudInit Version         |                                                   |
+-------------------------------------------------------------------------------+
| Config Guest Id           | sles15_64Guest                                    |
+-------------------------------------------------------------------------------+
| Hardware Version          | vmx-20                                            |
+-------------------------------------------------------------------------------+
| GuestInfo Guest Id        | sles15_64Guest                                    |
+-------------------------------------------------------------------------------+
| GuestInfo Guest Full Name | SUSE Linux Enterprise 15 (64-bit)                 |
+-------------------------------------------------------------------------------+
| GuestInfo Guest Family    | linuxGuest                                        |
+-------------------------------------------------------------------------------+
| GuestInfo Detailed Data   | bitness='64'                                      |
|                           | distroName='SLED'                                 |
|                           | distroVersion='15.3'                              |
|                           | familyName='Linux'                                |
|                           | kernelVersion='5.3.18-57-default'                 |
|                           | prettyName='SUSE Linux Enterprise Desktop 15 SP3' |
+-------------------------------------------------------------------------------+


Test Results (Total: 28, Failed: 0, No Run: 4, Elapsed Time: 01:43:30):
+-------------------------------------------------------------+
| Name                                 |   Status | Exec Time |
+-------------------------------------------------------------+
| deploy_vm_efi_nvme_vmxnet3           |   Passed | 00:25:07  |
| ovt_verify_install                   |   Passed | 00:06:55  |
| ovt_verify_status                    |   Passed | 00:02:21  |
| vgauth_check_service                 |   Passed | 00:00:37  |
| check_ip_address                     |   Passed | 00:00:39  |
| stat_balloon                         |   Passed | 00:00:36  |
| stat_hosttime                        |   Passed | 00:00:30  |
| check_inbox_driver                   |   Passed | 00:00:40  |
| check_os_fullname                    |   Passed | 00:00:30  |
| check_efi_firmware                   |   Passed | 00:00:30  |
| cpu_hot_add_basic                    |   Passed | 00:04:26  |
| check_quiesce_snapshot_custom_script |   Passed | 00:01:20  |
| memory_hot_add_basic                 |   Passed | 00:05:01  |
| device_list                          |   Passed | 00:02:31  |
| secureboot_enable_disable            |   Passed | 00:04:06  |
| e1000e_network_device_ops            |   Passed | 00:01:01  |
| vmxnet3_network_device_ops           |   Passed | 00:00:57  |
| cpu_multicores_per_socket            |   Passed | 00:09:42  |
| gosc_perl_dhcp                       | * No Run | 00:00:29  |
| gosc_perl_staticip                   | * No Run | 00:00:29  |
| gosc_cloudinit_dhcp                  | * No Run | 00:00:29  |
| gosc_cloudinit_staticip              | * No Run | 00:00:29  |
| paravirtual_vhba_device_ops          |   Passed | 00:02:00  |
| lsilogic_vhba_device_ops             |   Passed | 00:02:09  |
| lsilogicsas_vhba_device_ops          |   Passed | 00:01:56  |
| sata_vhba_device_ops                 |   Passed | 00:02:21  |
| nvme_vhba_device_ops                 |   Passed | 00:02:03  |
| ovt_verify_uninstall                 |   Passed | 00:02:21  |
+-------------------------------------------------------------+
```
```
+---------------------------------------------------------------+
| Guest OS Type             | Ubuntu 21.10 x86_64               |
+---------------------------------------------------------------+
| VMTools Version           | 11.3.0.29534 (build-18090558)     |
+---------------------------------------------------------------+
| CloudInit Version         | 21.3-1-g6803368d-0ubuntu3         |
+---------------------------------------------------------------+
| Config Guest Id           | ubuntu64Guest                     |
+---------------------------------------------------------------+
| Hardware Version          | vmx-20                            |
+---------------------------------------------------------------+
| GuestInfo Guest Id        | ubuntu64Guest                     |
+---------------------------------------------------------------+
| GuestInfo Guest Full Name | Ubuntu Linux (64-bit)             |
+---------------------------------------------------------------+
| GuestInfo Guest Family    | linuxGuest                        |
+---------------------------------------------------------------+
| GuestInfo Detailed Data   | architecture='X86'                |
|                           | bitness='64'                      |
|                           | distroName='Ubuntu'               |
|                           | distroVersion='21.10'             |
|                           | familyName='Linux'                |
|                           | kernelVersion='5.13.0-19-generic' |
|                           | prettyName='Ubuntu 21.10'         |
+---------------------------------------------------------------+


Test Results (Total: 28, Failed: 1, No Run: 2, Elapsed Time: 01:41:54):
+-------------------------------------------------------------+
| Name                                 |   Status | Exec Time |
+-------------------------------------------------------------+
| deploy_ubuntu_ova                    |   Passed | 00:03:56  |
| ovt_verify_install                   |   Passed | 00:04:38  |
| ovt_verify_status                    |   Passed | 00:01:00  |
| vgauth_check_service                 |   Passed | 00:01:04  |
| check_ip_address                     |   Passed | 00:00:56  |
| stat_balloon                         |   Passed | 00:01:00  |
| stat_hosttime                        |   Passed | 00:00:59  |
| check_inbox_driver                   |   Passed | 00:01:07  |
| check_os_fullname                    |   Passed | 00:00:53  |
| check_efi_firmware                   | * No Run | 00:01:05  |
| cpu_hot_add_basic                    |   Passed | 00:04:26  |
| check_quiesce_snapshot_custom_script |   Passed | 00:01:46  |
| memory_hot_add_basic                 |   Passed | 00:05:40  |
| device_list                          |   Passed | 00:02:56  |
| secureboot_enable_disable            | * No Run | 00:00:52  |
| e1000e_network_device_ops            |   Passed | 00:01:52  |
| vmxnet3_network_device_ops           |   Passed | 00:01:48  |
| cpu_multicores_per_socket            |   Passed | 00:09:20  |
| gosc_perl_dhcp                       |   Passed | 00:05:13  |
| gosc_perl_staticip                   |   Passed | 00:04:30  |
| gosc_cloudinit_dhcp                  | * Failed | 00:04:50  |
| gosc_cloudinit_staticip              |   Passed | 00:04:35  |
| paravirtual_vhba_device_ops          |   Passed | 00:02:23  |
| lsilogic_vhba_device_ops             |   Passed | 00:02:46  |
| lsilogicsas_vhba_device_ops          |   Passed | 00:02:19  |
| sata_vhba_device_ops                 |   Passed | 00:02:35  |
| nvme_vhba_device_ops                 |   Passed | 00:02:20  |
| ovt_verify_uninstall                 |   Passed | 00:02:41  |
+-------------------------------------------------------------+
```
```
+---------------------------------------------------------------------------------------------+
| Guest OS Type             | Flatcar 3033.2.0 x86_64                                         |
+---------------------------------------------------------------------------------------------+
| VMTools Version           | 11.3.5.31214 (build-18557794)                                   |
+---------------------------------------------------------------------------------------------+
| CloudInit Version         |                                                                 |
+---------------------------------------------------------------------------------------------+
| Config Guest Id           | other3xLinux64Guest                                             |
+---------------------------------------------------------------------------------------------+
| Hardware Version          | vmx-20                                                          |
+---------------------------------------------------------------------------------------------+
| GuestInfo Guest Id        | other5xLinux64Guest                                             |
+---------------------------------------------------------------------------------------------+
| GuestInfo Guest Full Name | Other 5.x Linux (64-bit)                                        |
+---------------------------------------------------------------------------------------------+
| GuestInfo Guest Family    | linuxGuest                                                      |
+---------------------------------------------------------------------------------------------+
| GuestInfo Detailed Data   | architecture='X86'                                              |
|                           | bitness='64'                                                    |
|                           | buildNumber='2021-12-10-1820'                                   |
|                           | distroName='Flatcar Container Linux by Kinvolk'                 |
|                           | distroVersion='3033.2.0'                                        |
|                           | familyName='Linux'                                              |
|                           | kernelVersion='5.10.84-flatcar'                                 |
|                           | prettyName='Flatcar Container Linux by Kinvolk 3033.2.0 (Oklo)' |
+---------------------------------------------------------------------------------------------+


Test Results (Total: 28, Failed: 2, No Run: 9, Elapsed Time: 01:19:30):
+-------------------------------------------------------------+
| Name                                 |   Status | Exec Time |
+-------------------------------------------------------------+
| deploy_flatcar_ova                   |   Passed | 00:04:11  |
| ovt_verify_install                   | * No Run | 00:00:35  |
| ovt_verify_status                    |   Passed | 00:01:05  |
| vgauth_check_service                 | * No Run | 00:00:30  |
| check_ip_address                     | * Failed | 00:01:31  |
| stat_balloon                         |   Passed | 00:00:32  |
| stat_hosttime                        |   Passed | 00:00:29  |
| check_inbox_driver                   |   Passed | 00:00:35  |
| check_os_fullname                    |   Passed | 00:00:37  |
| check_efi_firmware                   | * No Run | 00:00:35  |
| cpu_hot_add_basic                    |   Passed | 00:02:29  |
| check_quiesce_snapshot_custom_script |   Passed | 00:00:46  |
| memory_hot_add_basic                 |   Passed | 00:05:05  |
| device_list                          |   Passed | 00:02:57  |
| secureboot_enable_disable            | * No Run | 00:00:32  |
| e1000e_network_device_ops            |   Passed | 00:00:55  |
| vmxnet3_network_device_ops           |   Passed | 00:01:01  |
| cpu_multicores_per_socket            |   Passed | 00:09:30  |
| gosc_perl_dhcp                       | * No Run | 00:00:31  |
| gosc_perl_staticip                   | * No Run | 00:00:26  |
| gosc_cloudinit_dhcp                  | * No Run | 00:00:31  |
| gosc_cloudinit_staticip              | * No Run | 00:00:28  |
| paravirtual_vhba_device_ops          |   Passed | 00:02:12  |
| lsilogic_vhba_device_ops             | * Failed | 00:02:32  |
| lsilogicsas_vhba_device_ops          |   Passed | 00:02:41  |
| sata_vhba_device_ops                 |   Passed | 00:02:21  |
| nvme_vhba_device_ops                 |   Passed | 00:02:09  |
| ovt_verify_uninstall                 | * No Run | 00:00:33  |
+-------------------------------------------------------------+
```

**Comment by keirazhang** [comment]
```
  - 'GUEST OS: Microsoft Windows Server 2022 Datacenter'
  - 'Results: PASS (19/22), FAIL (3/22) in 02:32:03.'
  - ' * deploy_vm_efi_nvme_e1000e.............................................Passed'
  - ' * secureboot_enable_disable.............................................Passed'
  - ' * wintools_complete_install_verify......................................Passed'
  - ' * check_efi_firmware....................................................Passed'
  - ' * check_ip_address......................................................Passed'
  - ' * check_os_fullname.....................................................Passed'
  - ' * mouse_driver_vmtools..................................................Passed'
  - ' * vgauth_check_service..................................................Passed'
  - ' * stat_balloon..........................................................Passed'
  - ' * paravirtual_vhba_device_ops...........................................Passed'
  - ' * lsilogicsas_vhba_device_ops...........................................Passed'
  - ' * sata_vhba_device_ops..................................................Failed'
  - ' * nvme_vhba_device_ops..................................................Passed'
  - ' * e1000e_network_device_ops.............................................Passed'
  - ' * vmxnet3_network_device_ops............................................Passed'
  - ' * memory_hot_add_basic..................................................Passed'
  - ' * check_quiesce_snapshot................................................Failed'
  - ' * cpu_multicores_per_socket.............................................Passed'
  - ' * gosc_sanity_staticip..................................................Passed'
  - ' * gosc_sanity_dhcp......................................................Passed'
  - ' * wintools_uninstall_verify.............................................Failed'
  - ' * cpu_hot_add_basic.....................................................Passed'
```
```
  - 'GUEST OS: Microsoft Windows 10 Enterprise 10.0.19044.0'
  - 'Results: PASS (22/22) in 03:17:37.'
  - ' * deploy_vm_efi_nvme_e1000e.............................................Passed'
  - ' * secureboot_enable_disable.............................................Passed'
  - ' * wintools_complete_install_verify......................................Passed'
  - ' * check_efi_firmware....................................................Passed'
  - ' * check_ip_address......................................................Passed'
  - ' * check_os_fullname.....................................................Passed'
  - ' * mouse_driver_vmtools..................................................Passed'
  - ' * vgauth_check_service..................................................Passed'
  - ' * stat_balloon..........................................................Passed'
  - ' * paravirtual_vhba_device_ops...........................................Passed'
  - ' * lsilogicsas_vhba_device_ops...........................................Passed'
  - ' * sata_vhba_device_ops..................................................Passed'
  - ' * nvme_vhba_device_ops..................................................Passed'
  - ' * e1000e_network_device_ops.............................................Passed'
  - ' * vmxnet3_network_device_ops............................................Passed'
  - ' * memory_hot_add_basic..................................................Passed'
  - ' * check_quiesce_snapshot................................................Passed'
  - ' * cpu_multicores_per_socket.............................................Passed'
  - ' * gosc_sanity_staticip..................................................Passed'
  - ' * gosc_sanity_dhcp......................................................Passed'
  - ' * wintools_uninstall_verify.............................................Passed'
  - ' * cpu_hot_add_basic.....................................................Passed'
```
```
  - 'GUEST OS: Microsoft Windows 10 Enterprise 10.0.19044.0'
  - 'Results: PASS (21/22), FAIL (1/22) in 03:33:58.'
  - ' * deploy_vm_efi_nvme_e1000e.............................................Passed'
  - ' * secureboot_enable_disable.............................................Passed'
  - ' * wintools_complete_install_verify......................................Passed'
  - ' * check_efi_firmware....................................................Passed'
  - ' * check_ip_address......................................................Passed'
  - ' * check_os_fullname.....................................................Passed'
  - ' * mouse_driver_vmtools..................................................Passed'
  - ' * vgauth_check_service..................................................Passed'
  - ' * stat_balloon..........................................................Passed'
  - ' * paravirtual_vhba_device_ops...........................................Passed'
  - ' * lsilogicsas_vhba_device_ops...........................................Passed'
  - ' * sata_vhba_device_ops..................................................Passed'
  - ' * nvme_vhba_device_ops..................................................Passed'
  - ' * e1000e_network_device_ops.............................................Passed'
  - ' * vmxnet3_network_device_ops............................................Passed'
  - ' * memory_hot_add_basic..................................................Passed'
  - ' * check_quiesce_snapshot................................................Failed'
  - ' * cpu_multicores_per_socket.............................................Passed'
  - ' * gosc_sanity_staticip..................................................Passed'
  - ' * gosc_sanity_dhcp......................................................Passed'
  - ' * wintools_uninstall_verify.............................................Passed'
  - ' * cpu_hot_add_basic.....................................................Passed'
```
```
  - 'GUEST OS: Microsoft Windows 11 Enterprise 10.0.22000.0'
  - 'Results: PASS (20/22), FAIL (2/22) in 04:32:03.'
  - ' * deploy_vm_efi_nvme_e1000e.............................................Passed'
  - ' * secureboot_enable_disable.............................................Passed'
  - ' * wintools_complete_install_verify......................................Passed'
  - ' * check_efi_firmware....................................................Passed'
  - ' * check_ip_address......................................................Passed'
  - ' * check_os_fullname.....................................................Passed'
  - ' * mouse_driver_vmtools..................................................Passed'
  - ' * vgauth_check_service..................................................Passed'
  - ' * stat_balloon..........................................................Passed'
  - ' * paravirtual_vhba_device_ops...........................................Passed'
  - ' * lsilogicsas_vhba_device_ops...........................................Passed'
  - ' * sata_vhba_device_ops..................................................Passed'
  - ' * nvme_vhba_device_ops..................................................Passed'
  - ' * e1000e_network_device_ops.............................................Passed'
  - ' * vmxnet3_network_device_ops............................................Passed'
  - ' * memory_hot_add_basic..................................................Passed'
  - ' * check_quiesce_snapshot................................................Passed'
  - ' * cpu_multicores_per_socket.............................................Passed'
  - ' * gosc_sanity_staticip..................................................Failed'
  - ' * gosc_sanity_dhcp......................................................Failed'
  - ' * wintools_uninstall_verify.............................................Passed'
  - ' * cpu_hot_add_basic.....................................................Passed'
 ```

File: linux/deploy_vm/wait_autoinstall_completed.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Copyright 2022 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+- name: "Wait for autoinstall completed message"
```
Comment (Tomorrow9) [inline linux/deploy_vm/wait_autoinstall_completed.yml:4]: Our purpose is reducing and not use tasks delegate_to ESXi host, how about use URL like this "https://xx.xx.xx.xx/folder/test_file.log?dsName=datastore2&dcPath=ha-datacenter" to get the file content?

File: linux/deploy_vm/rebuild_unattend_install_iso.yml [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+# Rebuild OS ISO image with unattended install file
+# Parameter:
+#   rebuilt_unattend_iso_path: Local path to the rebuilt ISO image with unattend install config file
+#
+- name: "Set fact of the absolute path of source ISO file on ESXi server"
+  set_fact:
+    iso_file_path_on_esxi: "{{ os_installation_iso_list[0] | replace('[', '/vmfs/volumes/') | replace('] ', '/') }}"
+
+# Fetch ISO file from ESXi datastore
+- name: "Download ISO image from ESXi server datastore"
```
Comment (Tomorrow9) [inline linux/deploy_vm/rebuild_unattend_install_iso.yml:13]: And this:
Our purpose is reducing and not use tasks delegate_to ESXi host, how about use URL like this "https://xx.xx.xx.xx/folder/test_file.log?dsName=datastore2&dcPath=ha-datacenter" to download the file?

File: linux/deploy_vm/rebuild_unattend_install_iso.yml [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+          with_items: "{{ boot_menu_cfg_file_paths }}"
+
+        - name: "Set facts for rebuilding ISO options"
+          set_fact:
+            unattend_iso_label: "-V {{ src_iso_file_name | splitext | first }}"
+            boot_info_opts: "-R -l -L -D -b isolinux/isolinux.bin -c isolinux/boot.cat -no-emul-boot -boot-load-size 4 -boot-info-table -eltorito-alt-boot --eltorito-boot boot/grub2/efiboot.img -no-emul-boot"
+
+        # Create the unattended install ISO image
+        - name: "Rebuild ISO image with unattend install config file"
+          shell: "xorriso -as mkisofs {{ boot_info_opts }} {{ unattend_iso_label }} . > {{ rebuilt_unattend_iso_path }}"
```
Comment (Tomorrow9) [inline linux/deploy_vm/rebuild_unattend_install_iso.yml:103]: Does this "xorriso" package always available and need to be added in requirements.txt?

**Review by Tomorrow9 — CHANGES_REQUESTED** [review]

File: linux/deploy_vm/rebuild_unattend_install_iso.yml (hunk 3 above)
Comment (keirazhang) [reply linux/deploy_vm/rebuild_unattend_install_iso.yml:103]: Yes, it is required. I just updated requirements.txt. Once the PR merged, we need to update the docker image as well.

File: linux/deploy_vm/wait_autoinstall_completed.yml (hunk 1 above)
Comment (keirazhang) [reply linux/deploy_vm/wait_autoinstall_completed.yml]: `common/vm_wait_logbundle_msg.yml` was added to check or wait for a message in VM's log bundles.

File: linux/deploy_vm/rebuild_unattend_install_iso.yml (hunk 2 above)
Comment (keirazhang) [reply linux/deploy_vm/rebuild_unattend_install_iso.yml]: `common/esxi_download_datastore_file.yml` was added to download datastore file from URL.

File: common/esxi_download_datastore_file.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+- include_tasks: esxi_check_delete_datastore_file.yml
+  vars:
+    file_in_datastore_ops: file
+    file_in_datastore: "{{ src_datastore }}"
+    file_in_datastore_path: "{{ src_file_path }}"
+    file_in_datastore_failed_ignore: False
+
+- name: "Set fact of downloading URL for the src datastore file"
+  set_fact:
+    datastore_file_url: "{{ ds_file_result.url }}"
```
Comment (Tomorrow9) [inline common/esxi_download_datastore_file.yml:20]: `ds_file_result.url` can be checked if it's defined or `datastore_file_url` will be set to "".

**Review by Tomorrow9 — CHANGES_REQUESTED** [review]

File: common/esxi_download_datastore_file.yml (hunk 4 above)
Comment (keirazhang) [reply common/esxi_download_datastore_file.yml:27]: Added an assert to check ds_file_result.url before set_fact.

**Review by Tomorrow9 — APPROVED** [review]
LGTM, thanks.

**Pull request merged**